### PR TITLE
functors task (1) - Pysarenko

### PR DESCRIPTION
**Pull request by Elzianor**

### CODE REVIEW

File: functors_task_1/student/include/impl.h [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
 /**
  * @todo Implement functor-generator that will return next beer brand (cyclic)
  * @example HoeGaarden -> Corona -> Carlsberg -> Bud -> ZlataPraha -> Leffe -> HoeGaarden...
  */
 struct BeerOrganizer
 {
+    BeerBrand operator()();
+
+private:
+    size_t currentIndex{1};
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:18]: avoid magic numbers

File: functors_task_1/student/include/impl.h [hunk 2]
```diff
@@ -16,16 +24,15 @@ struct BeerOrganizer
...
-{
-}
+bool isExpensiveBeer(const BeerBrand& brand);
 
 /**
  * @todo Implement lambda beer country equality comparator
  * @return true if beer county is the same, false otherwise
  */
-auto sameCountry = [](/**???*/)
+auto sameCountry = [](const BeerBrand& lhs, const BeerBrand& rhs)
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:33]: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const

File: functors_task_1/student/src/impl.cpp [hunk 3]
```diff
@@ -1,2 +1,29 @@
 #include "impl.h"
 
+BeerBrand BeerOrganizer::operator()() {
+    if (static_cast<size_t>(BeerBrand::Max) == currentIndex)
+        currentIndex = 1;
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:5]: avoid magic numbers, specify "from" and "to" const variables

File: functors_task_1/student/src/impl.cpp [hunk 4]
```diff
@@ -1,2 +1,29 @@
...
+
+    return static_cast<BeerBrand>(currentIndex++);
+}
+
+bool isExpensiveBeer(const BeerBrand& brand)
+{
+    std::vector<BeerBrand> expensive{BeerBrand::HoeGaarden, BeerBrand::Corona};
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:12]: 1. Why do you need dynamic container?
2. Why not the plain logical expression?

File: functors_task_1/student/src/impl.cpp [hunk 5]
```diff
@@ -1,2 +1,29 @@
...
+
+    return std::find(expensive.begin(), expensive.end(), brand) != std::end(expensive);
+}
+
+Cocktail MixingPolicy::mix(AlcoholDrink alcohol, NonAlcoholDrink nonAlcohol)
+{
+    using Recipe = std::tuple<AlcoholDrink, NonAlcoholDrink>;
+    using Ingredients = std::pair<AlcoholDrink, NonAlcoholDrink>;
+
+    std::map<Recipe, Cocktail> cookbook{{Ingredients(AlcoholDrink::Gin, NonAlcoholDrink::LimeJuice), Cocktail::Gimlet},
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:22]: 1. In such cases, unordered_map should be used, as you don't need the order
2. Why you need to recreate this object every time function is called?
3. Mark it as const

File: functors_task_1/student/src/impl.cpp [hunk 6]
```diff
@@ -1,2 +1,29 @@
...
+bool isExpensiveBeer(const BeerBrand& brand)
+{
+    std::vector<BeerBrand> expensive{BeerBrand::HoeGaarden, BeerBrand::Corona};
+
+    return std::find(expensive.begin(), expensive.end(), brand) != std::end(expensive);
+}
+
+Cocktail MixingPolicy::mix(AlcoholDrink alcohol, NonAlcoholDrink nonAlcohol)
+{
+    using Recipe = std::tuple<AlcoholDrink, NonAlcoholDrink>;
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:19]: why tuple and not the pair?

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: functors_task_1/student/include/impl.h (hunk 1 above)
Comment (Elzianor) [reply functors_task_1/student/include/impl.h]: Changed to use enum value

File: functors_task_1/student/include/impl.h (hunk 2 above)
Comment (Elzianor) [reply functors_task_1/student/include/impl.h]: Const and & removed.

File: functors_task_1/student/src/impl.cpp (hunk 3 above)
Comment (Elzianor) [reply functors_task_1/student/src/impl.cpp]: Changed to use enum values

File: functors_task_1/student/src/impl.cpp (hunk 4 above)
Comment (Elzianor) [reply functors_task_1/student/src/impl.cpp]: The purpose was to reuse existing find algorithm, but in this case it seems to be an overhead. Changed to switch - case.

File: functors_task_1/student/src/impl.cpp (hunk 5 above)
Comment (Elzianor) [reply functors_task_1/student/src/impl.cpp]: Changed to static const unordered_map.

File: functors_task_1/student/src/impl.cpp (hunk 6 above)
Comment (Elzianor) [reply functors_task_1/student/src/impl.cpp]: Tuple was removed, now only pair is used.

File: functors_task_1/student/src/impl.cpp [hunk 7]
```diff
@@ -2,28 +2,44 @@
...
+bool isExpensiveBeer(BeerBrand brand)
 {
-    std::vector<BeerBrand> expensive{BeerBrand::HoeGaarden, BeerBrand::Corona};
-
-    return std::find(expensive.begin(), expensive.end(), brand) != std::end(expensive);
+    switch(brand)
+    {
+    case BeerBrand::HoeGaarden:
+    case BeerBrand::Corona:
+        return true;
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:16]: avoid explicit true/false in your code

File: functors_task_1/student/src/impl.cpp [hunk 8]
```diff
@@ -2,28 +2,44 @@
...
+    using Recipe = std::pair<AlcoholDrink, NonAlcoholDrink>;
+
+    struct hashPair
+    {
+        size_t operator()(const Recipe& recipe) const
+        {
+            size_t first = static_cast<size_t>(recipe.first);
+            size_t second = static_cast<size_t>(recipe.second);
+
+            size_t cantorPairing = (first + second) * (first + second + 1) / 2 + second;
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:33]: prefer std::hash

File: functors_task_1/student/src/impl.cpp [hunk 9]
```diff
@@ -2,28 +2,44 @@
...
+    }
 }
 
 Cocktail MixingPolicy::mix(AlcoholDrink alcohol, NonAlcoholDrink nonAlcohol)
 {
-    using Recipe = std::tuple<AlcoholDrink, NonAlcoholDrink>;
-    using Ingredients = std::pair<AlcoholDrink, NonAlcoholDrink>;
+    using Recipe = std::pair<AlcoholDrink, NonAlcoholDrink>;
+
+    struct hashPair
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:26]: extend std namespace with hash specialization

File: functors_task_1/student/include/impl.h [hunk 10]
```diff
@@ -15,7 +15,7 @@ struct BeerOrganizer
     BeerBrand operator()();
 
 private:
-    size_t currentIndex{1};
+    size_t currentIndex{static_cast<size_t>(BeerBrand::HoeGaarden)};
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:18]: please specify range "from" and "to"

File: functors_task_1/student/src/impl.cpp [hunk 11]
```diff
@@ -2,28 +2,44 @@
 
 BeerBrand BeerOrganizer::operator()() {
     if (static_cast<size_t>(BeerBrand::Max) == currentIndex)
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:4]: extract value to a named const variable

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: functors_task_1/student/src/impl.cpp (hunk 7 above)
Comment (Elzianor) [reply functors_task_1/student/src/impl.cpp:16]: Could you please explain why?

File: functors_task_1/student/src/impl.cpp (hunk 8 above)
Comment (Elzianor) [reply functors_task_1/student/src/impl.cpp:33]: Xor is giving the same result in cases of (a xor b) and (b xor a).
So for example pairs (OrangeJuice = 1, Wine = 2) and (GrapefruitJuice = 2, Whiskey = 1) will result in the same hash value which is wrong. Cantor pairing approach on the other hand handles such cases correctly.

File: functors_task_1/student/src/impl.cpp (hunk 7 above)
Comment (smay1613) [reply functors_task_1/student/src/impl.cpp:16]: Doing so avoids verbosity and eliminates some opportunities for mistakes. Helps make style consistent and conventional.

File: functors_task_1/student/src/impl.cpp (hunk 8 above)
Comment (smay1613) [reply functors_task_1/student/src/impl.cpp:33]: no problem, but make it more explicit - extract to some function, I've suggested xor for simplicity

**Pull request closed without merging**